infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java

```diff
@@ -141,280 +158,315 @@ private boolean isBootstrapped(CorfuMsg msg, ChannelHandlerContext ctx, IServerR
         return true;
     }
 
+    private boolean isBootstrapped(HeaderMsg reqHeader, ChannelHandlerContext ctx, IServerRouter r) {
+        if(getCurrentLayout() == null) {
+            r.sendNoBootstrapError(reqHeader, ctx);
+            return false;
+        }
+
+        return true;
+    }
+
     // Helper Methods
 
     /**
      * Handle a layout request message.
      *
-     * @param msg              corfu message containing LAYOUT_REQUEST
+     * @param req              corfu message containing LAYOUT_REQUEST
      * @param ctx              netty ChannelHandlerContext
      * @param r                server router
      */
-    @ServerHandler(type = CorfuMsgType.LAYOUT_REQUEST)
-    public synchronized void handleMessageLayoutRequest(CorfuPayloadMsg<Long> msg,
-                                                    ChannelHandlerContext ctx, IServerRouter r) {
-        if (!isBootstrapped(msg, ctx, r)) {
-            return;
-        }
+    @RequestHandler(type = RequestPayloadMsg.PayloadCase.LAYOUT_REQUEST)
+    public synchronized void handleLayoutRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        if(!isBootstrapped(req.getHeader(), ctx, r)) return;
 
-        final long msgEpoch = msg.getPayload();
+        final long payloadEpoch = req.getPayload().getLayoutRequest().getEpoch();
         final long serverEpoch = getServerEpoch();
 
-        if (msgEpoch <= serverEpoch) {
-            r.sendResponse(ctx, msg, new LayoutMsg(getCurrentLayout(), CorfuMsgType
-                    .LAYOUT_RESPONSE));
+        if(payloadEpoch <= serverEpoch) {
+            HeaderMsg responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            ResponseMsg response = getResponseMsg(responseHeader, getLayoutResponseMsg(getCurrentLayout()));
+            r.sendResponse(response, ctx);
         } else {
-            // else the client is somehow ahead of the server.
-            //TODO figure out a strategy to deal with this situation
-            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, serverEpoch));
-            log.warn("handleMessageLayoutRequest: Message Epoch {} ahead of Server epoch {}",
-                    msgEpoch, serverEpoch);
+            //TODO: Client is ahead of the server. Is any other handling required?
```

Comment:
You can remove the TODO this scenario can naturally happen. Imagine you run Paxos Layout at epoch 2. Its possible that out of the 3 nodes, only 2 nodes accepted epoch two layout. So, when a client requests a layout from the third node (i.e., the one that didn't participate in the paxos round, then it should just return the older layout. If you return a wrong epoch exception its fine, but superfluous because the clients need to check that they never regress the epoch the layout that they are tracking.